src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -67,69 +64,270 @@ public void WriteMessage(HubMessage message, Stream output)
+                                            try
+                                            {
+                                                var paramTypes = binder.GetParameterTypes(target);
+                                                arguments = BindArguments(reader, paramTypes);
+                                            }
+                                            catch (Exception ex)
+                                            {
+                                                argumentBindingException = ExceptionDispatchInfo.Capture(ex);
+                                            }
+                                        }
+                                        break;
+                                    case HeadersPropertyName:
+                                        JsonUtils.CheckRead(reader);
+                                        headers = ReadHeaders(reader);
+                                        break;
+                                    default:
+                                        // Skip read the property name
+                                        JsonUtils.CheckRead(reader);
+                                        // Skip the value for this property
+                                        reader.Skip();
+                                        break;
+                                }
+                                break;
+                            case JsonToken.EndObject:
+                                completed = true;
+                                break;
+                            default:
+                                break;
+                        }
+                    }
+                    while (reader.Read() && !completed);
 
-                    switch (type)
+                    if (!completed)
                     {
-                        case HubProtocolConstants.InvocationMessageType:
-                            return BindInvocationMessage(json, binder);
-                        case HubProtocolConstants.StreamInvocationMessageType:
-                            return BindStreamInvocationMessage(json, binder);
-                        case HubProtocolConstants.StreamItemMessageType:
-                            return BindStreamItemMessage(json, binder);
-                        case HubProtocolConstants.CompletionMessageType:
-                            return BindCompletionMessage(json, binder);
-                        case HubProtocolConstants.CancelInvocationMessageType:
-                            return BindCancelInvocationMessage(json);
-                        case HubProtocolConstants.PingMessageType:
-                            return PingMessage.Instance;
-                        default:
-                            throw new InvalidDataException($"Unknown message type: {type}");
+                        // This should throw if we stopped parsing before we found EndObject
+                        JsonUtils.CheckRead(reader);
```

Comment:
Make CheckRead return a bool (which will always be true) and use on line 212. Should be able to remove this if check.

Reply:
These nits are killing me.

Reply:
Just throw directly from the loop condition is what you mean right?